### PR TITLE
feat (app): Enable support for exporting xpubs to host

### PR DESCRIPTION
**Pull request by ujjwal-cyph**
Fixes https://app.clickup.com/t/9002019994/PRF-4813

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-4813 Add account support](https://app.clickup.com/t/9002019994/PRF-4813)

**Comment by ujjwal-cyph** [comment]
Depends on https://github.com/Cypherock/cypherock-common/pull/18

File: common/coin_support/coin_utils.h [hunk 1]
```diff
@@ -352,6 +351,21 @@ void derive_hdnode_from_path(const uint32_t *path,
...
+ * @param path
+ * @param path_length
+ * @param s_node
+ * @param hdnode
+ */
+bool derive_hdnode_from_node(const uint32_t *path,
+                             size_t path_length,
+                             const HDNode *s_node,
+                             HDNode *hdnode);
+
```
Comment (amanCypherock) [inline common/coin_support/coin_utils.h:368]: Would this function be required as a global function or can it be a static one?

File: common/coin_support/coin_utils.h [hunk 2]
```diff
@@ -321,10 +321,9 @@ int64_t byte_array_to_recv_txn_data(Receive_Transaction_Data *txn_data_ptr,
  *
  * @note
  */
-void generate_xpub(const uint32_t *path,
-                   const size_t path_length,
-                   const char *curve,
-                   const uint8_t *seed,
+bool generate_xpub(const uint32_t *path,
```
Comment (amanCypherock) [inline common/coin_support/coin_utils.h:324]: Update doxygen comments.

File: apps/btc_app/get_xpub.c [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+
+  return true;
+}
+
+static bool validate_request_data(btc_get_xpubs_request_t *request) {
+  bool status = true;
+
+  // TODO: enable checking if wallet is operational
+  // status = wallet_is_operational(request->initiate.wallet_id);
+  if (0 == request->initiate.derivation_paths_count) {
```
Comment (kartikCypherock) [inline apps/btc_app/get_xpub.c:206]: Send error to host

File: common/coin_support/coin_utils.c [hunk 4]
```diff
@@ -334,22 +334,29 @@ int64_t byte_array_to_add_coin_data(Add_Coin_Data *data_ptr,
...
-  hdnode_private_ckd(&t_node, path[path_length - 1]);
+  if (0 == hdnode_private_ckd(&t_node, path[path_length - 1])) {
+    // hdnode_private_ckd returns 1 when the derivation succeeds
+    status &= false;
+  }
   hdnode_fill_public_key(&t_node);
 
   get_version(path[0], path[1], NULL, &version);
-  hdnode_serialize_public(&t_node, fingerprint, version, str, 113);
+  if (0 == hdnode_serialize_public(&t_node, fingerprint, version, str, 113)) {
```
Comment (kartikCypherock) [inline common/coin_support/coin_utils.c:355]: Use macro to convey size of string buffer

File: common/coin_support/coin_utils.h (hunk 1 above)
Comment (ujjwal-cyph) [reply common/coin_support/coin_utils.h:368]: this will be removed altogether

File: common/coin_support/coin_utils.c (hunk 4 above)
Comment (ujjwal-cyph) [reply common/coin_support/coin_utils.c]: changed

File: common/coin_support/coin_utils.h (hunk 2 above)
Comment (ujjwal-cyph) [reply common/coin_support/coin_utils.h:321]: added

File: common/coin_support/coin_utils.h (hunk 1 above)
Comment (ujjwal-cyph) [reply common/coin_support/coin_utils.h]: removed

File: apps/btc_app/get_xpub.c [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+  wallet_name = (const char *)get_wallet_name(wallet_index);
+  // Take user consent to export xpub for the wallet
+  if (!confirm_with_user(wallet_name)) {
+    return;
+  }
+  core_status_set_flow_status(BTC_GET_XPUBS_STATUS_CONFIRM);
+
+  // TODO: call the reconstruct flow and get master node from the core
+  HDNode master_node = {0};    // = generate_master_node();
+  char xpub_list[init_req->derivation_paths_count][XPUB_SIZE];
```
[the remark this answers is not in the document]
Comment (ujjwal-cyph) [reply apps/btc_app/get_xpub.c]: changed

File: apps/btc_app/get_xpub.c (hunk 3 above)
Comment (ujjwal-cyph) [reply apps/btc_app/get_xpub.c:206]: changed

**Review by amanCypherock — APPROVED** [review]

**Review by kartikCypherock — APPROVED** [review]

**Pull request merged**